src/mongocxx/options/client_session.hpp

```diff
@@ -75,6 +96,7 @@ class MONGOCXX_API client_session {
 
    private:
     bool _causal_consistency = true;
+    bool _enable_snapshot_reads = false;
```

Comment:
The default handling behavior may require distinguishing whether `snapshot` and `causal_consistency` were unset by the user.

The [Causal Consistency Specification](https://github.com/mongodb/specifications/blob/c48d7f416e51dce8fd4c8d05bdb05790508ca3bb/source/causal-consistency/causal-consistency.rst#causalconsistency) notes:

> Note that the causalConsistency property is optional. The default value of this property is not supplied. If no value is supplied for causalConsistency the value will be inherited. Currently it is inherited from the global default which is defined to be true. In the future it might be inherited from client settings.

The [Snapshot Sessions Specification](https://github.com/mongodb/specifications/blob/c48d7f416e51dce8fd4c8d05bdb05790508ca3bb/source/sessions/snapshot-sessions.rst#snapshot) notes:

> Note that the snapshot property is optional. The default value of this property is false.
> Snapshot reads and causal consistency are mutually exclusive. Therefore if snapshot is set to true, causalConsistency must be false. Client MUST throw an error if both snapshot and causalConsistency are set to true. Snapshot reads are supported on both primaries and secondaries. 

Currently this `mongocxx::options::client_session` does not distinguish between `causal_consistency` being unset and `true`. The following causes an error:

```
auto client = mongocxx::client(mongocxx::uri());
auto opts = mongocxx::options::client_session();
opts.snapshot(true);
client.start_session(opts); // exception due to both snapshot and causal_consistency being set.
```

I suggest storing both as `bsoncxx::stdx::optional`.

Reply:
So... this is a question! Causal consistency is currently implemented, but is /not/ implemented as optional. Should that change here? And isn't the "default value... false" at odds with snapshot being optional? (Is there an inheritable "not supplied" setting?)

Reply:
> Causal consistency is currently implemented, but is /not/ implemented as optional. Should that change here?

I think so. Before `snapshot` was added, it did not matter that `causal_consistency` being unset and `true` were not distinguished. In both cases the result was a session with `causal_consistency` enabled.

But if `snapshot` is true:
- it is an error if `causal_consistency` is explicitly set to true
- otherwise, the default of `causal_consistency` evaluates to false.

> And isn't the "default value... false" at odds with snapshot being optional?

I am think of it in terms of "what the user sets" vs. "what the option evaluates to when start_session is called":


Set snapshot | Set causal_consistency | Evaluated snapshot | Evaluated causal_consistency
-- | -- | -- | --
unset | unset | FALSE | TRUE
unset | FALSE | FALSE | FALSE
unset | TRUE | FALSE | TRUE
FALSE | unset | FALSE | TRUE
FALSE | FALSE | FALSE | FALSE
FALSE | TRUE | FALSE | TRUE
TRUE | unset | TRUE | FALSE
TRUE | FALSE | TRUE | FALSE
TRUE | TRUE | error | error

Reply:
Unfortunately, in our current implementation from what I can see we don't actually get updated values from the server: they remain in the client-side structure. I'll see if there's a way to get a new, updated set via another call, which will enable us to test this.

Reply:
...there isn't really a good way to test this that isn't "after the fact", but I'm now satisfied that incorrect combinations will be (from the point of view of the client library/driver) be detected without allowing a valid connection.